app/lib/analyzer/backend.dart

```diff
@@ -203,27 +203,50 @@ class AnalysisBackend {
     return true;
   }
 
-  Future deleteObsoleteAnalysisEntries() async {
+  /// Deletes the obsolete [Analysis] instances from Datastore. An instance is
+  /// obsolete, if:
+  /// - it is **not** the latest for the pana version, or
+  /// - it is older than 6 months (except if it is the latest one).
+  Future deleteObsoleteAnalysis(String package, String version) async {
     final DateTime threshold =
         new DateTime.now().toUtc().subtract(obsoleteThreshold);
-    final Query query = db.query(Analysis)
-      ..order('timestamp')
-      ..filter('timestamp <', threshold);
-    await for (Analysis analysis in query.run()) {
-      final PackageVersionAnalysis pva =
-          (await db.lookup([analysis.parentKey])).single;
-      if (pva.latestAnalysis != analysis.analysis) {
-        _logger.info('Deleting obsolete Analysis: ${analysis.packageName} '
-            '${analysis.packageVersion} ${analysis.analysis}');
-        await db.withTransaction((Transaction tx) async {
-          final List list = await tx.lookup([analysis.key]);
-          if (list[0] != null) {
-            tx.queueMutations(deletes: [list[0].key]);
-          }
-          await tx.commit();
-        });
+    final Query scanQuery = db.query(
+      Analysis,
+      ancestorKey: db.emptyKey
+          .append(PackageAnalysis, id: package)
+          .append(PackageVersionAnalysis, id: version),
+    );
+    final List<Key> obsoleteKeys = <Key>[];
+
+    final List<_AnalysisRef> refs = <_AnalysisRef>[];
+    await for (Analysis analysis in scanQuery.run()) {
+      refs.add(new _AnalysisRef(
+          analysis.key, analysis.panaVersion, analysis.timestamp));
+    }
+    refs.sort((a, b) => a.timestamp.compareTo(b.timestamp));
+
+    // (length - 1) makes sure that we keep at least the latest version
+    for (int i = 0; i < refs.length - 1; i++) {
```

Comment:
Maybe add a comment here: `NOTE: This is an O(n^2) algorithm`.

Or much better: Use a `Map<String, Analysis> panaVersion2LatestAnalysis = {}`. Then you can just walk through the `refs` / `existingAnalysis` **once** update the `panaVersion2LatestAnalysis` and every time you replace an entry you put it into `List<Key> toBeDeleted`.